common/src/View/MapDocument.cpp

```diff
@@ -133,10 +118,79 @@
 #include <sstream>
 #include <string>
 #include <type_traits>
+#include <unordered_map>
 #include <vector>
 
 namespace TrenchBroom {
     namespace View {
+        template <typename N, typename L>
+        static std::optional<std::vector<std::pair<Model::Node*, Model::NodeContents>>> applyToNodeContents(const std::vector<N*>& nodes, L lambda) {
```

Comment:
This was quite confusing at first glance. I'd add docs with the following:

```
L needs two overloads:

bool operator()(Model::Entity&);
bool operator()(Model::Brush&);

and should modify the given entity/brush in place. If any lambda invocation returns false the entire `applyToNodeContents` invocation returns std::nullopt.
```

Reply:
Also I'm not convinced about making N a template param. This is just so you can do 

```
std::vector<EntityNode*> entityNodes;
applyToNodeContents(entityNodes, ...);
```
without having to do a vec_element_cast to `std::vector<Node*>`?

It's going to cause binary size bloat + compile time bloat to make it a template param.

Reply:
Good point about the documentation: 82b133ee7efba6e204dfe34f99bdae7a147b52bb and 7c22d70948afd9ca5c6bb75a7b1a14012b052653

Reply:
Regarding the template / code bloat: A new template will be instantiated for each invocation anyway because each invocation passes a different lambda. I am going to investigate how we can reduce the compile time for `MapDocument` in a separate issue.